crates/api/src/instance.rs

```diff
@@ -72,6 +72,35 @@ pub struct Instance {
 }
 
 impl Instance {
+    /// Creates a new [`Instance`] from the previously compiled [`Module`] and
+    /// list of `imports` specified, similar to `Instance::new`, and performs
+    /// [WASI ABI initialization]:
+    ///  - If the module is a command, the `_start` function is run and `None`
+    ///    is returned.
+    ///  - If the module is a reactor, the `_initialize` function is run and
+    ///    the initialized `Instance` is returned.
+    ///
+    /// [WASI ABI initialization]: https://github.com/WebAssembly/WASI/blob/master/design/application-abi.md#current-unstable-abi
+    pub fn new_wasi_abi(module: &Module, imports: &[Extern]) -> Result<Option<Instance>, Error> {
```

Comment:
I think this API makes sense given WASI today, but this feels sort of uncomfortable. This means that we're splitting the world between "instantiate this module" and "instantiate this wasi module". These constructor semantics need to be bubbled to all callers and such. 

Would it be possible to avoid splitting into two worlds? One option might be to have this instead of a constructor:

```rust
impl Instance {
    fn initialize_wasi(self) -> Result<Option<Instance>, Error> {
        // ...
    }
}
```

That way you'd always create the `Instance` the same way and it's up to you what to do with it after that.

Reply:
I'm imagining `new_wasi_abi` will be a common case. The name is perhaps confusing because this doesn't mean "run with all the WASI APIs available", but just "run the WASI startup functions if present". I actually wonder if we should rename this to `new` and rename the current `new` to `new_raw` or something, to emphasize that it's not automatically running the WASI startup functions (though it does still run the wasm start function).

Reply:
I personally still feel that this is not the API we want. Adding a new constructor for instances which "does the wasi thing" is infectious and will continue to force us to duplicate API surface area into the future. I feel remembering to call `new_wasi_abi` is basically equivalent to remembering to call `initialize_wasi`, and in terms of composability I feel that an explicit function to initialize is more convenient.

Reply:
[A lucet change](https://github.com/bytecodealliance/lucet/pull/506) to disconnect initialization from instantiation also points in this direction.

I experimented with a few different API approaches to this, and came up with the patch that's now pushed here. With this, the Rust API's `Instance::new` and `Linker::instantiate` return a `NewInstance`, which is a wrapper around `Instance` where the programmer has to initialize it before getting the `Instance` out. We may need to do something different for the C API though.